distributed/utils.py

```diff
@@ -1000,3 +1001,49 @@ def __init__(self, *args, **kwargs):
 
     def emit(self, record):
         self.deque.append(record)
+
+
+class ThrottledGC(object):
+    """Wrap gc.collect to protect against excessively repeated calls.
+
+    Allows to run throttled garbage collection in the workers as a
+    countermeasure to e.g.: https://github.com/dask/zict/issues/19
+
+    collect() does nothing when repeated calls are so costly and so frequent
+    that the thread would spend more than max_in_gc_frac doing GC.
+
+    warn_if_longer is a duration in seconds (10s by default) that can be used
+    to control to log a warning level message whenever an actual call to
+    gc.collect() lasts too long.
+    """
+    def __init__(self, max_in_gc_frac=0.2, warn_if_longer=10, logger=None):
+        self.max_in_gc_frac = max_in_gc_frac
+        self.warn_if_longer = 10
+        self.last_collect = time()
+        self.last_gc_duration = 0
+        self.logger = logger if logger is not None else _logger
+
+    def collect(self):
+        # In case of non-monotonicity in the clock, assume that any Python
+        # duration last at least 1 microsecond.
+        collect_start = time()
+        elapsed = max(collect_start - self.last_collect, 1e-6)
+        if self.last_gc_duration / elapsed < self.max_in_gc_frac:
+            self.logger.debug("Calling gc.collect(). %0.3fs elapsed since "
+                              "previous call.", elapsed)
+            gc.collect()
+            self.last_collect = collect_start
+            self.last_gc_duration = max(time() - collect_start, 1e-6)
+            if self.last_gc_duration > self.warn_if_longer:
+                self.logger.warning("gc.collect() took %0.3fs. This is usually"
```

Comment:
Hmm, I'm curious why you think this is useful. Do you know a use case where that warning would have helped?

Reply:
To warn datascientists that are inadvertently processing a list of 100 million small dicts at a time in each task instead of processing more smaller lists of a couple of million each instead. For instance parsing large JSON files in large chunks instead of small chunks. 

I assume that large GC pauses might make the worker non-responsive. But if you think this is useless we can get rid of this warning.

Reply:
> I assume that large GC pauses might make the worker non-responsive.

It will. But what if using many objects is deliberate and required by the workload? Do we want to warn against any possible programming mistake?

Reply:
I think that users would appreciate getting a warning if GC is taking seconds.  They have responded well to similar feedback that we have provided in the past (like long event loop pauses).